serial/serialposix.py

```diff
@@ -674,9 +674,38 @@ def flush(self):
         """\
         Flush of file like objects. In this case, wait until all data
         is written.
+
+        On macOS with PTY devices, tcdrain() can block indefinitely due to
+        a known kernel issue. For PTYs, we skip tcdrain() since data is
+        already in the kernel buffer after write().
         """
         if not self.is_open:
             raise PortNotOpenError()
+
+        # On macOS, tcdrain() can block indefinitely on PTY devices.
+        # See: https://github.com/pyserial/pyserial/issues/625
+        #      https://github.com/python/cpython/issues/97001
+        if sys.platform.startswith('darwin'):
+            try:
+                dev_name = os.ttyname(self.fd)
+                # macOS PTY slave devices are /dev/ttys*
+                # BSD-style PTY devices are /dev/pty[pqrs]*
+                is_pty = (dev_name.startswith('/dev/ttys') or
+                          dev_name.startswith('/dev/ptyp') or
+                          dev_name.startswith('/dev/ptyq') or
+                          dev_name.startswith('/dev/ptyr') or
+                          dev_name.startswith('/dev/ptys'))
+                if is_pty:
+                    # For PTY devices on macOS, tcdrain() blocks indefinitely
+                    # because it waits for data to be read from the other end,
+                    # which may never happen. Since the data is already in the
+                    # kernel buffer after write(), we consider flush() complete.
+                    return
+            except (OSError, IOError):
```

Comment:
`IOError` is an alias for `OSError`.

```suggestion
            except OSError:
```